### ISSUE TITLE
[jit] support `isinstance(foo, dict)`

### ISSUE THREAD

**Issue by suo** [pytorch/pytorch#37516]
Today doing something like:

```
@torch.jit.script
def foo(s: Any):
    if isinstance(s, dict):
        print(s.items())
```
gives:
```
RuntimeError: Unknown type name 'dict':
```
And capitalizing "`dict`" will produce a segfault. Yikes!

A related issue is that doing something like:
```
isinstance(foo, Dict[str, Tensor])
```
is not actually possible in Python, since the type annotations are mostly static constructs that the interpreter doesn't know anything about.

A potential compromise is to have `isinstance(foo, dict)` refine the type of `foo` to `Dict[Any, Any]`. And then the user can further type refine to get an actual things they want. So:
```
if isinstance(foo, dict):
    for k, v in foo.items():
        assert isinstance(k, str)
        assert isinstance(v, Tensor)
```

Another potential solution is to have a `torch.jit.isinstance(foo, Dict[str, tensor])` that boils down to `isinstance(foo, dict)` in Python but performs the correct type refinement in TS.

cc @suo @gmagogsfm

**Comment by jansel** [comment]
Not clear how to solve this, need to come up with design and run it by others.

**Comment by edqwerty10** [comment]
Hi pytorch team, just a question, are there plans to also implement a similar check with `list`, ie. `isinstance(foo, list)`.
Currently, the following code breaks in script mode:
**Code Snippet:**
```
import torch
from typing import Any

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if isinstance(input, list):
            for el in input:
                print(el)

scripted_module = torch.jit.script(TestModule())
scripted_module(["1", "2"])
```
**Output**
```
'list' object is not iterable:
  File "<ipython-input-4-1d53cd51275f>", line 13
    def forward(self, input: Any) -> None:
        if isinstance(input, list):
            for el in input:
            ~~~~~~~~~~~~~~~~
                print(el)
                ~~~~~~~~ <--- HERE
```

**Comment by yf225** [comment]
This is important for avoiding code duplication in PyPer models. Consider the following use case:
```python
import torch
from typing import Any, List

class TestModule(torch.nn.Module):
    def __init__(self):
        super(TestModule, self).__init__()
    
    def call(self, input1: str, input2: str) -> str:
        return input1

    def forward(self, input: Any) -> None:
        if torch.jit.is_scripting() and isinstance(input, List[str]):
            # This can be a big code block in actual use cases
            for el in input:
                print(el)
        elif not torch.jit.is_scripting() and isinstance(input, list):
            # Code duplication!
            for el in input:
                print(el)
        else:
            raise Exception

module = TestModule()
scripted_module = torch.jit.script(module)

module(["1", "2"])
scripted_module(["1", "2"])
```
One theoretically plausible workaround which doesn't work:
```python
def forward(self, input: Any) -> None:
    if (torch.jit.is_scripting() and isinstance(input, List[str])) or (
        not torch.jit.is_scripting() and isinstance(input, list)
    ):  # Unfortunately this doesn't work in scripting...
        for el in input:
            print(el)
```
So to avoid code duplication, we need `isinstance(foo, dict)` and `isinstance(foo, list)` to work. :)

**Comment by gmagogsfm** [comment]
I am pretty sure I am missing something, but would it be enough to encapsulate the logic into functions to avoid duplication? 

To be clear, even with all code in a function, this is still inconvenient and we should improve the situation, just wanted to get more clarity. Thanks.

**Comment by gmagogsfm** [comment]
How do we solve the type covariance problem if we treat `isinstance(foo, Dict)` as `isinstance(foo, Dict[Any, Any])`? Without changing covariance rules, this would never evaluate to `True`. Therefore, it is really not a viable solution.

Creating `torch.jit.isinstance` is suboptimal as it requires users to change their code, but it does sidestep this problem. Thoughts?

**Comment by gmagogsfm** [comment]
Pointers to implementing `torch.jit.isinstance` if we were to take that route.

1. Add `torch.jit.isinstance` in python layer. In eager mode, it checks whether given object is `Dict`, then check all of its keys/values for subtype. 

2. For scripting, a `SpecialFormValue` should be generated to represent a call to `torch.jit.isinstance`. It should be created here:
https://github.com/pytorch/pytorch/blob/cf48872d28f945d47793f63e19c54dd15bf580f7/torch/csrc/jit/python/python_sugared_value.cpp#L918

3. `torch.jit.isinstance` should ultimately translate to something like `prim::isinstance(foo, DictType[TensorType, StrType]`, we already support this operator.

Essentially, we are enhancing Python's `isinstance`, which doesn't support subtypes, by providing this API so that users can write same code for both eager and scripting mode.
 
Optionally but ideally, we should also support other container types like dict, optional, tuple.

**Comment by gmagogsfm** [comment]
Hi @edqwerty10 @yf225,

@lilyjjo just added new API to meet your need. Can you try replace all your `isinstance` calls with `torch.jit.isinstance`, which should support checking types like `Dict[str, Tensor]` and behave the same in eager and JIT mode. Let us know if it works out for you.

**Comment by esqu1** [comment]
@gmagogsfm  @Lilyjjo It seems that this is giving different behaviors between whether the module subclasses ScriptModule or not:

```
import torch
class TestModule(torch.jit.ScriptModule):
    def __init__(self):
        super().__init__()
        
    def forward(self, x: Any) -> torch.Tensor:
        if torch.jit.isinstance(x, Dict[int, torch.Tensor]):
            return torch.tensor([2])
        else:
            return torch.tensor([3])
    
m = TestModule()
m({2: torch.tensor([2])})
```
will return `torch.Tensor([2])`, but
```
import torch
class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        
    def forward(self, x: Any) -> torch.Tensor:
        if torch.jit.isinstance(x, Dict[int, torch.Tensor]):
            return torch.tensor([2])
        else:
            return torch.tensor([3])
    
m = TestModule()
m2 = torch.jit.script(m)
m2({2: torch.tensor([2])})
```
will return `torch.Tensor([3])`. Are you able to reproduce this on your end?

**Comment by Lilyjjo** [comment]
@esqu1 I can reproduce the issue, thank you for reporting it. I don't think the issue is caused by what is being subclassed. When I run the torch.jit.ScriptModule version of your repro the code is running in eager mode (which is a separate issue). The root cause of the issue I think is that torch.jit.isinstance's scripted version is failing to resolve the Dict[int, torch.Tensor] type properly while the eager version can handle it. I'll start to debug this

**Comment by Lilyjjo** [comment]
The difference between eager and scripting should be fixed by #46130
